src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java

```diff
@@ -0,0 +1,356 @@
+ *     net.snowflake.ingest.streaming.internal.ParquetRowBuffer.ParquetChunkData})
+ */
+public abstract class AbstractRowBuffer<T> implements RowBuffer<T> {
+  private static final Logging logger = new Logging(AbstractRowBuffer.class);
+
+  // Snowflake table column logical type
+  enum ColumnLogicalType {
+    ANY,
+    BOOLEAN(1),
+    ROWINDEX,
+    NULL(15),
+    REAL(8),
+    FIXED(2),
+    TEXT(9),
+    CHAR,
+    BINARY(10),
+    DATE(7),
+    TIME(6),
+    TIMESTAMP_LTZ(3),
+    TIMESTAMP_NTZ(4),
+    TIMESTAMP_TZ(5),
+    INTERVAL,
+    RAW,
+    ARRAY(13, true),
+    OBJECT(12, true),
+    VARIANT(11, true),
+    ROW,
+    SEQUENCE,
+    FUNCTION,
+    USER_DEFINED_TYPE,
+    ;
+
+    private final int ordinal;
+    private final boolean object;
+
+    ColumnLogicalType() {
+      this(-1);
+    }
+
+    ColumnLogicalType(int ordinal) {
+      this(ordinal, false);
+    }
+
+    ColumnLogicalType(int ordinal, boolean object) {
+      this.ordinal = ordinal;
+      this.object = object;
+    }
+
+    public int getOrdinal() {
+      return ordinal;
+    }
+
+    public boolean isObject() {
+      return object;
+    }
+  }
+
+  // Snowflake table column physical type
+  enum ColumnPhysicalType {
+    ROWINDEX(9),
+    DOUBLE(7),
+    SB1(1),
+    SB2(2),
+    SB4(3),
+    SB8(4),
+    SB16(5),
+    LOB(8),
+    BINARY,
+    ROW(10),
+    ;
+
+    private final int ordinal;
+
+    ColumnPhysicalType() {
+      this.ordinal = -1;
+    }
+
+    ColumnPhysicalType(int ordinal) {
+      this.ordinal = ordinal;
+    }
+
+    public int getOrdinal() {
+      return ordinal;
+    }
+  }
+
+  // Map the column name to the stats
+  @VisibleForTesting Map<String, RowBufferStats> statsMap;
+
+  // Temp stats map to use until all the rows are validated
+  @VisibleForTesting Map<String, RowBufferStats> tempStatsMap;
+
+  // Lock used to protect the buffers from concurrent read/write
+  final Lock flushLock;
+
+  // Current row count
+  @VisibleForTesting volatile int rowCount;
+
+  // Current buffer size
+  volatile float bufferSize;
+
+  // Reference to the Streaming Ingest channel that owns this buffer
+  @VisibleForTesting final SnowflakeStreamingIngestChannelInternal<T> owningChannel;
+
+  // Names of non-nullable columns
+  final Set<String> nonNullableFieldNames;
+
+  AbstractRowBuffer(SnowflakeStreamingIngestChannelInternal<T> channel) {
+    this.owningChannel = channel;
+    this.nonNullableFieldNames = new HashSet<>();
+    this.flushLock = new ReentrantLock();
+    this.rowCount = 0;
+    this.bufferSize = 0F;
+
+    // Initialize empty stats
+    this.statsMap = new HashMap<>();
+    this.tempStatsMap = new HashMap<>();
+  }
+
+  /**
+   * Get the current buffer size
+   *
+   * @return the current buffer size
+   */
+  @Override
+  public float getSize() {
+    return this.bufferSize;
+  }
+
+  /**
+   * Verify that the input row columns are all valid
```

Comment:
Let's specify what `valid` means in this context. Should we also compare data / logical types here as well?